### PR TITLE
[Enhancement] Apply query queue based on slot estimation 

### PR DESCRIPTION
**Pull request by ZiheLiu**
## Why I'm doing:

Apply query queue based on slot estimation

### Implementation

The slot resource is divided into small extra slots and normal slots.

- Small slots are tried to allocate slots from the small extra slots. If it fails, will try to allocate them from the normal slots. The small slot is defined as the query that requires only one slot.

- Selecting slots to allocate from the normal slots is based on the smooth weighted round-robin algorithm. The slots are divided into multiple sub-queues according to the number of slots required by the query. Each sub-queue has a different weight. 
    - The implementation of the smooth weighted round-robin algorithm.
        - Besides, if the peaking sub-queue is empty, will update `state` as usual and record the skip times of this sub-queue. Repeat this process until the peaking sub-queue is not empty. When this sub-queue is not empty anymore, will give it some compensations according to the skip times.
        - The rules to divide sub-queues are as follows:
            - The minimum slots of `subQueues[i]` is $2^{b-(7-i)}$, that is, the minimum slots of 0~7 sub-queues are $2^{b-7}$, $2^{b-6}$, $2^{b-5}$, $2^{b-4}$, $2^{b-3}$, $2^{b-2}$, $2^{b-1}$, $2^b$, respectively. where $b$ denotes `lastSubQueueLog2Bound`.
            - The weight of `subQueues[i]` is $2.5^{7-i}$, that is, the weights of 0~7 sub-queues are $2.5^7$, $2.5^6$, $2.5^5$, $2.5^4$, $2.5^3$, $2.5^2$, $2.5^1$, $2.5^0$ respectively.

### How to use

The query queue strategy uses the previous strategy by default. To switch to the new strategy, add `enable_query_queue_v2=true` to `fe.conf` and **restart all the FEs**.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/metric/MetricWithLabelGroup.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+    private final Map<String, E> labelValueToMetric = Maps.newConcurrentMap();
+
+    public MetricWithLabelGroup(String labelKey, Supplier<E> metricCreator) {
+        this.labelKey = labelKey;
+        this.metricCreator = metricCreator;
+    }
+
+    public E getMetric(String labelValue) {
+        return labelValueToMetric.computeIfAbsent(labelValue, k -> {
```
[the remark this answers is not in the document]
Comment (ZiheLiu) [reply fe/fe-core/src/main/java/com/starrocks/metric/MetricWithLabelGroup.java:34]: Needn't according to the document of `ConcurrencyHashMap`:
> If the specified key is not already associated with a value, attempts to compute its value using the given mapping function and enters it into this map unless null.  **The entire method invocation is performed atomically, so the function is applied at most once per key**.  Some attempted update operations on this map by other threads may be blocked while computation is in progress, so the computation should be short and simple, and must not attempt to update any other mappings of this map.

File: fe/fe-core/src/main/java/com/starrocks/plugin/AuditEvent.java [hunk 2]
```diff
@@ -121,6 +121,8 @@ public enum EventType {
     public double planMemCosts = -1;
     @AuditField(value = "PendingTimeMs")
     public long pendingTimeMs = -1;
+    @AuditField(value = "Slots")
```
[the remark this answers is not in the document]
Comment (ZiheLiu) [reply fe/fe-core/src/main/java/com/starrocks/plugin/AuditEvent.java:124]: Let me check later.

File: fe/fe-core/src/main/java/com/starrocks/qe/QueryQueueManager.java [hunk 3]
```diff
@@ -123,11 +126,26 @@ private LogicalSlot createSlot(DefaultCoordinator coord) throws UserException {
...
+        return new LogicalSlot(coord.getQueryId(), frontend.getNodeName(), groupId, numSlots, expiredPendingTimeMs,
                 expiredAllocatedTimeMs, frontend.getStartTime(), numFragments, pipelineDop);
     }
+
+    private int estimateNumSlots(ConnectContext context, DefaultCoordinator coord) {
+        QueryQueueOptions opts = QueryQueueOptions.createFromEnvAndQuery(coord);
+
+        SlotEstimator estimator = SlotEstimatorFactory.create(opts);
+        int numSlots = estimator.estimateSlots(opts, context, coord);
+        if (opts.isEnableQueryQueueV2()) {
```
[the remark this answers is not in the document]
Comment (ZiheLiu) [reply fe/fe-core/src/main/java/com/starrocks/qe/QueryQueueManager.java:144]: It is to write `Slots` to audit log.

File: fe/fe-core/src/main/java/com/starrocks/common/Config.java [hunk 4]
```diff
@@ -664,6 +664,15 @@ public class Config extends ConfigBase {
     @ConfField
     public static int http_port = 8030;
 
+    @ConfField
+    public static boolean enable_query_queue_v2 = false;
+    @ConfField(mutable = true)
+    public static int query_queue_v2_concurrency_level = 4;
+    @ConfField(mutable = true)
+    public static int query_queue_v2_num_rows_per_slot = 4096;
```
[the remark this answers is not in the document]
Comment (ZiheLiu) [reply fe/fe-core/src/main/java/com/starrocks/common/Config.java:672]: - `cpuCosts` comes from audit log.
- `num_rows` comes from explain costs.

[the remark this answers is not in the document]
Comment (ZiheLiu) [reply fe/fe-core/src/main/java/com/starrocks/common/Config.java:672]: Let me add some comments.

File: fe/fe-core/src/main/java/com/starrocks/qe/QueryQueueManager.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ZiheLiu) [reply fe/fe-core/src/main/java/com/starrocks/qe/QueryQueueManager.java:144]: Let me add some comments.

**Review by satanson — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Review by HangyuanLiu — APPROVED** [review]

**Pull request merged**